hexl/fft/inv-fft-avx512.cpp

```diff
@@ -0,0 +1,366 @@
+
+    // Weights
+    __m512d v_W_real = _mm512_set1_pd(*W_1C_intrlvd++);
+    __m512d v_W_imag = _mm512_set1_pd(*W_1C_intrlvd++);
+
+    // assume 8 | t
+    for (size_t j = 0; j < gap; j += 16) {
+      __m512d v_X_real = _mm512_loadu_pd(v_X_pt_real);
+      __m512d v_X_imag = _mm512_loadu_pd(v_X_pt_imag);
+
+      __m512d v_Y_real = _mm512_loadu_pd(v_Y_pt_real);
+      __m512d v_Y_imag = _mm512_loadu_pd(v_Y_pt_imag);
+
+      ComplexInvButterfly(&v_X_real, &v_X_imag, &v_Y_real, &v_Y_imag, v_W_real,
+                          v_W_imag);
+
+      _mm512_storeu_pd(v_X_pt_real, v_X_real);
+      _mm512_storeu_pd(v_X_pt_imag, v_X_imag);
+
+      _mm512_storeu_pd(v_Y_pt_real, v_Y_real);
+      _mm512_storeu_pd(v_Y_pt_imag, v_Y_imag);
+
+      // Increase operand & result pointers
+      v_X_pt_real += 2;
+      v_X_pt_imag += 2;
+      v_Y_pt_real += 2;
+      v_Y_pt_imag += 2;
+    }
+    offset += (gap << 1);
+  }
+}
+
+// Takes operand as 8 complex interleaved: This is 8 real parts followed by
+// its 8 imaginary parts.
+// Returns operand as 1 complex interleaved: One real part followed by its
+// imaginary part.
+void ComplexFinalInvT8(double* operand_8C_intrlvd, const double* W_1C_intrlvd,
+                       uint64_t gap, uint64_t m,
+                       const double* scalar = nullptr) {
+  size_t offset = 0;
+
+  __m512d v_scalar;
+  if (scalar != nullptr) {
+    v_scalar = _mm512_set1_pd(*scalar);
+  }
+
+  HEXL_LOOP_UNROLL_4
+  for (size_t i = 0; i < (m >> 1); i++) {
+    // Referencing operand
+    double* X_real = operand_8C_intrlvd + offset;
+    double* X_imag = X_real + 8;
+
+    double* Y_real = X_real + gap;
+    double* Y_imag = X_imag + gap;
+
+    __m512d* v_X_pt_real = reinterpret_cast<__m512d*>(X_real);
+    __m512d* v_X_pt_imag = reinterpret_cast<__m512d*>(X_imag);
+
+    __m512d* v_Y_pt_real = reinterpret_cast<__m512d*>(Y_real);
+    __m512d* v_Y_pt_imag = reinterpret_cast<__m512d*>(Y_imag);
+
+    // Weights
+    __m512d v_W_real = _mm512_set1_pd(*W_1C_intrlvd++);
+    __m512d v_W_imag = _mm512_set1_pd(*W_1C_intrlvd++);
+
+    if (scalar != nullptr) {
+      v_W_real = _mm512_mul_pd(v_W_real, v_scalar);
+      v_W_imag = _mm512_mul_pd(v_W_imag, v_scalar);
+    }
+
+    // assume 8 | t
+    for (size_t j = 0; j < gap; j += 16) {
+      __m512d v_X_real = _mm512_loadu_pd(v_X_pt_real);
+      __m512d v_X_imag = _mm512_loadu_pd(v_X_pt_imag);
+      __m512d v_Y_real = _mm512_loadu_pd(v_Y_pt_real);
+      __m512d v_Y_imag = _mm512_loadu_pd(v_Y_pt_imag);
+
+      ComplexInvButterfly(&v_X_real, &v_X_imag, &v_Y_real, &v_Y_imag, v_W_real,
+                          v_W_imag, scalar);
+
+      ComplexWriteInvInterleavedT8(&v_X_real, &v_X_imag, &v_Y_real, &v_Y_imag,
+                                   v_X_pt_real, v_Y_pt_real);
+
+      // Increase operand & result pointers
+      v_X_pt_real += 2;
+      v_X_pt_imag += 2;
+      v_Y_pt_real += 2;
+      v_Y_pt_imag += 2;
+    }
+    offset += (gap << 1);
```

Comment:
I don't think gap is modified in this function. I am not sure if the compiler is therefore guaranteed to only compute (gap << 1) only once. It may be best to declare a variable in the for loop init section and update the offset in the increment for loop section.